### PR TITLE
Support opset-13 specs of controlflow ops (Loop, If)

### PR DESCRIPTION
**Pull request by hariharans29**
**Description**:

Support Loop and If.

Enhances the onnx test runner application to parse sequence tensors from serialized messages and enhances the test infrastructure by adding comparison logic for OrtValues containing sequence tensors

**Motivation and Context**
Make ORT opset-13 compliant

### CODE REVIEW

File: onnxruntime/core/providers/cpu/tensor/unsqueeze.cc [hunk 1]
```diff
@@ -31,12 +41,26 @@ Status UnsqueezeBase::PrepareCompute(OpKernelContext* ctx, Prepare& p) const {
   ORT_ENFORCE(X != nullptr);
   auto& input_tensor = *X;
 
-  // New dimension count is the current dimensions + the number of entries in axes_
+  std::vector<int64_t> axes;
```
Comment (hariharans29) [inline onnxruntime/core/providers/cpu/tensor/unsqueeze.cc:44]: Will wait for https://github.com/microsoft/onnxruntime/pull/5691 to go through and re-base with master after that change

File: onnxruntime/core/providers/cpu/tensor/unsqueeze.cc [hunk 2]
```diff
@@ -43,14 +43,15 @@ Status UnsqueezeBase::PrepareCompute(OpKernelContext* ctx, Prepare& p) const {
...
   std::vector<int64_t> axes;
   size_t num_inputs = ctx->InputCount();
-  if (num_inputs == 2) { //axes is an input
+  if (num_inputs == 2) {  //axes is an input
     const Tensor* axes_tensor = ctx->Input<Tensor>(1);
     ORT_ENFORCE(axes_tensor != nullptr, "Axes input is null");
-    ORT_ENFORCE(axes_tensor->Shape().NumDimensions() == 1,
-                "An axes tensor must be a vector tensor.");
-    auto nDims = static_cast<size_t>(axes_tensor->Shape()[0]);
+    ORT_ENFORCE(axes_tensor->Shape().NumDimensions() == 0 ||
```
Comment (hariharans29) [inline onnxruntime/core/providers/cpu/tensor/unsqueeze.cc:49]: A looser check is fine - the loop test uses a scalar to specify the axes for an unsqueeze node in its subgraph and the spec doesn't say that it is an error for a non 1-D case.

File: onnxruntime/test/onnx/TestCase.cc [hunk 3]
```diff
@@ -473,57 +494,105 @@ void OnnxTestCase::LoadTestData(size_t id, onnxruntime::test::HeapBuffer& b,
...
-    const ONNX_NAMESPACE::TensorProto& input = test_data_pbs[input_index];
-    size_t len = 0;
+  out.emplace(name_finalized, std::move(v1));
+}
+
+void OnnxTestCase::ConvertTestData(const ONNX_NAMESPACE::SequenceProto& test_data_pb,
+                                   onnxruntime::test::HeapBuffer& b,
+                                   bool is_input, size_t i,
+                                   std::unordered_map<std::string, Ort::Value>& out) const {
+  std::string name = test_data_pb.name();
```
Comment (pranavsharma) [inline onnxruntime/test/onnx/TestCase.cc:551]: nit: const ref

File: onnxruntime/core/providers/cpu/cpu_execution_provider.cc [hunk 4]
```diff
@@ -590,895 +592,897 @@ KernelCreateInfo BuildKernelCreateInfo<void>() {
 
 Status RegisterOnnxOperatorKernels(KernelRegistry& kernel_registry) {
   static const BuildKernelCreateInfoFn function_table[] = {
-    BuildKernelCreateInfo<void>,  //default entry to avoid the list become empty after ops-reducing
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/cpu_execution_provider.cc:593]: Can the diffs in this file be made relevant to this PR?

File: onnxruntime/core/framework/TensorSeq.h [hunk 5]
```diff
@@ -28,6 +28,9 @@ class TensorSeq {
   }
 
   void SetElements(std::vector<Tensor>&& tensors) {
+    // TODO:
+    // (1) Do we want to ensure elem_type is set before accepting tensors ?
```
Comment (pranavsharma) [inline onnxruntime/core/framework/TensorSeq.h:32]: I think the caller ensures this. We can instead document this function accordingly.

File: include/onnxruntime/core/session/onnxruntime_cxx_inline.h [hunk 6]
```diff
@@ -736,6 +736,19 @@ inline Value Value::CreateSequence(std::vector<Value>& values) {
...
+  return Value{out};
+}
+
+template <typename T>
+inline Value Value::CreateEmptyTensorSequence() {
+  OrtValue* out;
+  ThrowOnError(GetApi().CreateEmptyTensorSequence(TypeToTensorType<T>::type, &out));
+  return Value{out};
+}
+
```
Comment (skottmckay) [inline include/onnxruntime/core/session/onnxruntime_cxx_inline.h:751]: What's the use case where someone using the API needs to create an empty sequence?

File: onnxruntime/core/providers/cpu/cpu_execution_provider.cc (hunk 4 above)
Comment (skottmckay) [reply onnxruntime/core/providers/cpu/cpu_execution_provider.cc:593]: Diffs seem off. The previous indent of 2 looks more correct than the indent of 4 in the PR. 

---
In reply to: [518505101](https://github.com/microsoft/onnxruntime/pull/5665#discussion_r518505101) [](ancestors = 518505101)

Comment (hariharans29) [reply onnxruntime/core/providers/cpu/cpu_execution_provider.cc:593]: Sure- I don't know why it shows up like this - I ll fix it

File: onnxruntime/core/providers/cpu/controlflow/loop.cc [hunk 7]
```diff
@@ -20,6 +20,7 @@
 #include "core/framework/utils.h"
 #include "core/providers/cpu/tensor/utils.h"
 #include "core/framework/session_options.h"
+#include "core/framework//TensorSeq.h"
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/controlflow/loop.cc:24]: nit: //

File: include/onnxruntime/core/session/onnxruntime_cxx_inline.h (hunk 6 above)
Comment (hariharans29) [reply include/onnxruntime/core/session/onnxruntime_cxx_inline.h:751]: Apart from the node test case for opst-13 loop which consumes an empty sequence as a graph input (and populates the sequence every iteration), it does seem weird that currently the APIs only allow creation of sequences that have atleast one element (seems like a very strict constraint to arbitrarily impose on any user)

File: onnxruntime/core/providers/cpu/controlflow/loop.cc [hunk 8]
```diff
@@ -411,6 +438,7 @@ void LoopImpl::SaveOutputsAndUpdateFeeds(const std::vector<OrtValue>& last_outpu
 
   // save loop outputs as we have to concatenate at the end
   for (int j = info_.num_loop_carried_vars; j < info_.num_outputs; ++j) {
+    ORT_ENFORCE(last_outputs[j + 1].IsTensor(), "All scan outputs MUST be tensors");
     loop_output_tensors_[j - info_.num_loop_carried_vars].push_back(last_outputs[j + 1]);  // skip 'cond' in output
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/controlflow/loop.cc:442]: This is enforced by the onnx type constraint checking isn't it?

File: onnxruntime/core/providers/cpu/controlflow/loop.cc [hunk 9]
```diff
@@ -463,9 +491,31 @@ Status LoopImpl::Execute(const FeedsFetchesManager& ffm) {
...
+      TensorSeq* output = context_.Output<TensorSeq>(output_idx);
+      output->SetType(data.DataType());
+
+      AllocatorPtr alloc;
+      auto status = context_.GetTempSpaceAllocator(&alloc);
+      if (!status.IsOK()) {
+        ORT_THROW("Unable to get an allocator");
+      }
+      for (auto it = data.begin(); it != data.end(); ++it) {
+        Tensor tmp(it->DataType(), onnxruntime::TensorShape(it->Shape()), alloc);
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/controlflow/loop.cc:512]: nit:call data.end() once
`for (auto it = data.begin(), end = data.end(); it != end; ++it){`

File: onnxruntime/test/onnx/TestCase.cc [hunk 10]
```diff
@@ -403,21 +413,31 @@ static bool read_config_file(const std::basic_string<PATH_CHAR_TYPE>& path, std:
...
-    }
-    google::protobuf::io::FileInputStream f(tensor_fd, protobuf_block_size_in_bytes);
-    f.SetCloseOnDelete(true);
-    ONNX_NAMESPACE::TensorProto tensor;
-    if (!tensor.ParseFromZeroCopyStream(&f)) {
-      ORT_THROW("parse file '", ToMBString(pb_files.at(i)), "' failed");
-    }
-    input_pbs->emplace_back(tensor);
+static void LoadTensor(PATH_STRING_TYPE& pb_file, ONNX_NAMESPACE::TensorProto& input_pb) {
+  int tensor_fd;
```
Comment (skottmckay) [inline onnxruntime/test/onnx/TestCase.cc:417]: Can both these args be const? Same question for LoadSequenceTensor.

File: onnxruntime/test/onnx/TestCase.cc (hunk 3 above)
Comment (hariharans29) [reply onnxruntime/test/onnx/TestCase.cc:551]: Sure, but these diffs are misleading. This is the existing code as is. I can clean it up.

File: onnxruntime/test/onnx/TestCase.cc (hunk 10 above)
Comment (hariharans29) [reply onnxruntime/test/onnx/TestCase.cc:417]: Sure, but these diffs are misleading. This is the existing code as is. I can clean it up. (I didn't pay too much attention to already working code - I only spent time to adapt existing code )

File: include/onnxruntime/core/session/onnxruntime_cxx_inline.h (hunk 6 above)
Comment (skottmckay) [reply include/onnxruntime/core/session/onnxruntime_cxx_inline.h:751]: Not quite following. When does a user need to use the API to create an empty sequence? Seems like the model could provide a default empty sequence in a non-const initializer if that was needed.

8KB is a significant binary size increase for these changes, so we should trim everything we can. If we need a way to create an empty sequence (not convinced that we do yet), at the very least the templatized API method is not required.

File: onnxruntime/core/framework/data_types.cc [hunk 11]
```diff
@@ -971,6 +971,72 @@ const std::vector<MLDataType>& DataTypeImpl::AllNumericTensorTypes() {
   return all_numeric_size_tensor_types;
 }
 
+const std::vector<MLDataType>& DataTypeImpl::AllFixedSizeTensorAndSequenceTensorTypes() {
+  static std::vector<MLDataType> all_fixed_size_tensor_and_sequence_types =
```
Comment (skottmckay) [inline onnxruntime/core/framework/data_types.cc:975]: Can we define these using the existing ones? May reduce the binary size growth a little and avoid potential issues with inconsistent updates.

e.g.

```
static std::vector<MLDataType> all_fixed_size_tensor_and_sequence_types =
            [](){
              auto t = AllFixedSizeTensorTypes(); 
              const auto& s = AllSequenceTensorTypes();
              t.insert(t.end(), s.begin(), s.end());
              return t;
            }();
```

File: include/onnxruntime/core/session/onnxruntime_cxx_inline.h (hunk 6 above)
Comment (hariharans29) [reply include/onnxruntime/core/session/onnxruntime_cxx_inline.h:751]: When does a user need to use the API to create an empty sequence?
Whenever the user has a model that consumes an empty sequence. The test model was just a specific illustration. It scales to any model that consumes an empty sequence. Yes, the model could use a non-const initializer but that is kind of beside the point for me. It fixes this specific issue but doesn't fix the gap in the API. The API allows creation of sequences in general but only as long as they have atleast one element. Don't you think that is an arbitrary constraint given that ONNX itself allows creating empty sequences ?

Comment (pranavsharma) [reply include/onnxruntime/core/session/onnxruntime_cxx_inline.h:751]: I haven't seen a single real world model so far that uses tensor sequences, let alone models that take empty tensor sequences as inputs. Hence this wasn't supported before. Given the existence of CreateValue API that supports non-tensor types (non-empty), I would say let's think this more before committing it as a public API. The worst is you will not be able to run one artificial node test.

Comment (hariharans29) [reply include/onnxruntime/core/session/onnxruntime_cxx_inline.h:751]: This is the distribution of the 7~8 KB that this change adds:

image

As you can see the bulk of the size (> 5KB) comes from the op changes and other miscellaneous pieces needed. The API adds <  2KB and the value addition it provides is that it adds some completeness to sequence creation using the ORT API. It just feels weird that we allow creating sequences in general but we draw the line just before creating empty sequences. Having said that, we could pull out the API and save 2KB and move the node test to the filtered list as well and that is fine too.

File: onnxruntime/core/providers/cpu/cpu_execution_provider.cc (hunk 4 above)
Comment (hariharans29) [reply onnxruntime/core/providers/cpu/cpu_execution_provider.cc:593]: Fixed

File: onnxruntime/core/framework/TensorSeq.h (hunk 5 above)
Comment (hariharans29) [reply onnxruntime/core/framework/TensorSeq.h:32]: Documented in a comment

File: onnxruntime/core/providers/cpu/controlflow/loop.cc (hunk 7 above)
Comment (hariharans29) [reply onnxruntime/core/providers/cpu/controlflow/loop.cc:24]: Fixed

File: onnxruntime/core/providers/cpu/controlflow/loop.cc (hunk 8 above)
Comment (hariharans29) [reply onnxruntime/core/providers/cpu/controlflow/loop.cc:442]: Is there something in the framework that ensures that the type of the OrtValue will be the same as the ONNX inferred type ?

File: onnxruntime/test/onnx/TestCase.cc (hunk 3 above)
Comment (hariharans29) [reply onnxruntime/test/onnx/TestCase.cc:551]: Fixed

File: onnxruntime/test/onnx/TestCase.cc (hunk 10 above)
Comment (hariharans29) [reply onnxruntime/test/onnx/TestCase.cc:417]: The first argument can be const, I would have thought the second one can't be given that is the instance whose fields we will be setting as part of the proto message parsing action...

File: onnxruntime/core/framework/data_types.cc (hunk 11 above)
Comment (hariharans29) [reply onnxruntime/core/framework/data_types.cc:975]: Fixed for this and for existing types

File: include/onnxruntime/core/session/onnxruntime_cxx_inline.h (hunk 6 above)
Comment (hariharans29) [reply include/onnxruntime/core/session/onnxruntime_cxx_inline.h]: Pulled out the API + made changes in the test runner accordingly

**Review by pranavsharma — DISMISSED** [review]

**Review by skottmckay — DISMISSED** [review]
:shipit:

**Review by skottmckay — APPROVED** [review]

**Pull request merged**